### PR TITLE
Add option to list symbols in Object.getOwnPropertyNames()

### PR DESCRIPTION
**Pull request by pmarkee**
Also a little refactoring of Object.getOwnPropertyNames()

JerryScript-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/base/ecma-helpers-collection.c [hunk 1]
```diff
@@ -144,6 +144,31 @@ ecma_collection_push_back (ecma_collection_t *collection_p, /**< value collectio
...
+                        uint32_t index /**< index of the element to be removed */)
+{
+  JERRY_ASSERT (collection_p != NULL);
+  JERRY_ASSERT (collection_p->item_count > 0);
+  JERRY_ASSERT (index < collection_p->item_count);
+
+  ecma_value_t *buffer_p = collection_p->buffer_p;
+  const uint32_t len = collection_p->item_count;
+
+  ecma_free_value (buffer_p[index]);
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-collection.c]: That should be the caller's responsibility.

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 2]
```diff
@@ -783,13 +783,14 @@ ecma_builtin_object_object_assign (ecma_object_t *target_p, /**< target object *
     /* 5.b.iii */
     /* TODO: extends this collection if symbols will be supported */
     ecma_collection_t *props_p = ecma_op_object_get_property_names (from_obj_p, ECMA_LIST_CONVERT_FAST_ARRAYS
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: IMO the Todo comment above can be removed.

File: jerry-core/ecma/operations/ecma-objects.c [hunk 3]
```diff
@@ -2081,7 +2107,8 @@ ecma_op_object_get_property_names (ecma_object_t *obj_p, /**< object */
 
                  Also if is_symbols_only is true and prop_name is not symbol
                  we should skip the current property e.g. Object.getOwnPropertySymbols. */
-              if (JERRY_UNLIKELY (is_symbols_only != ecma_prop_name_is_symbol (name_p)))
+              bool is_symbol = ecma_prop_name_is_symbol (name_p);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-objects.c]: Please update the comment above.

File: jerry-core/ecma/operations/ecma-objects.c [hunk 4]
```diff
@@ -2112,6 +2139,32 @@ ecma_op_object_get_property_names (ecma_object_t *obj_p, /**< object */
 
               if (is_add)
               {
+                uint32_t index = ecma_string_get_array_index (name_p);
+                if (index != ECMA_STRING_NOT_ARRAY_INDEX)
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-objects.c]: NO local variable is needed.

File: jerry-core/ecma/operations/ecma-objects.c [hunk 5]
```diff
@@ -1969,6 +1972,28 @@ ecma_op_object_get_property_names (ecma_object_t *obj_p, /**< object */
...
+      uint32_t index = ecma_string_get_array_index (name_p);
+
+      if (index != ECMA_STRING_NOT_ARRAY_INDEX)
+      {
+        array_index_named_properties_count++;
+      }
+      else if (is_array_indices_only)
+      {
+        ecma_free_value (buffer_p[i]);
+        ecma_collection_remove (prop_names_p, i);
```
[the remark this answers is not in the document]
Comment (pmarkee) [reply jerry-core/ecma/operations/ecma-objects.c]: Good idea, this is not a very big change in these functions and then we don't need to add `ecma_collection_remove` at all.

**Comment by pmarkee** [comment]
Thanks for the review @rerobika , I updated the PR.

File: jerry-core/ecma/builtin-objects/ecma-builtins.c [hunk 6]
```diff
@@ -891,18 +891,16 @@ ecma_builtin_try_to_instantiate_property (ecma_object_t *object_p, /**< object *
  */
 void
 ecma_builtin_list_lazy_property_names (ecma_object_t *object_p, /**< a built-in object */
-                                       bool separate_enumerable, /**< true -  list enumerable properties into
-                                                                  *           main collection, and non-enumerable
-                                                                  *           to collection of 'skipped non-enumerable'
-                                                                  *           properties,
-                                                                  *   false - list all properties into main collection.
-                                                                  */
+                                       uint32_t opts, /**< listing options */
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtins.c]: Could we update the comment to reference the `ecma_list_properties_options_t` enum?
like: `uint32_t opts, /**< listing options using flags from ecma_list_properties_options_t */`

or maybe we should just mention it at the method's comments? What do you think?
(btw: this also applies to the other places where the `opts` is used)

**Comment by pmarkee** [comment]
Thanks for the review @galpeter . I have added a test case including symbols to the `Object.assign()` test, and a new test file to the `es2015` directory to check that `Object.getOwnPropertyNames()` does not include symbols by default.

**Pull request merged**